src/pytorch_lightning/strategies/colossalai.py

```diff
@@ -0,0 +1,294 @@
+            chunk_manager = ChunkManager(
+                chunk_size,
+                self.process_group,
+                self.enable_distributed_storage,
+                GeminiManager.get_default_device(self.placement_policy),
+            )
+            gemini_manager = GeminiManager(self.placement_policy, chunk_manager)
+            assert isinstance(self.model, (pl.LightningModule, _LightningPrecisionModuleWrapperBase))
+            model = _LightningModuleWrapperBase(self.model)
+            self.model = ZeroDDP(model, gemini_manager, self.force_outputs_fp32)
+            pl_module._colossalai_zero = [self.model]
+        else:
+            self.model = pl_module._colossalai_zero[0]
+        if is_training:
+            self.optimizers = [
+                ZeroOptimizer(optimizer, self.model, gpu_margin_mem_ratio=self.gpu_margin_mem_ratio, **self.amp_kwargs)
+            ]
+
+    def setup(self, trainer: "pl.Trainer") -> None:
+        if not isinstance(self.accelerator, CUDAAccelerator):
+            raise MisconfigurationException(
+                f"Colossalai is only supported on GPU but `{self.accelerator.__class__.__name__}` is used."
+            )
+        self.accelerator.setup(trainer)
+
+        assert self.lightning_module is not None
+        self.lightning_module._device = self.root_device
+
+        if trainer.accumulate_grad_batches > 1:
+            raise MisconfigurationException(
+                "Colossalai does not support gradient accumulation now. " "please set `accumulate_grad_batches` to 1."
+            )
+
+        accumulation_scheduler = trainer.accumulation_scheduler
+        if accumulation_scheduler.epochs != [0]:
+            raise MisconfigurationException(
+                "Colossalai currently does not support different `accumulate_grad_batches` at different epochs."
+            )
+        self.setup_optimizers(trainer)
+
+        if not isinstance(self.precision_plugin, ColossalAIPrecisionPlugin):
+            raise MisconfigurationException("`ColossalaiStrategy` is only compatible with `ColossalAIPrecisionPlugin`.")
+        self.setup_precision_plugin()
+        self.model_to_device()
+
+    @property
+    def root_device(self) -> torch.device:
+        if self.parallel_devices is not None:
+            return self.parallel_devices[self.local_rank]
+        return get_current_device()
+
+    def model_to_device(self) -> None:
+        pl_module = self.lightning_module
+        pl_module._device = self.root_device
+        for child in pl_module.modules():
+            if child is not pl_module and getattr(child, "_colossalai_module", None) is not True:
+                child.to(self.root_device)
+
+    @property
+    def lightning_module(self) -> Optional["pl.LightningModule"]:
+        if isinstance(self.model, ZeroDDP):
+            return self.model.module.lightning_module
+        return super().lightning_module
+
+    def teardown(self) -> None:
+        return
```

Comment:
it doesn't work with ddp.teardown?

Reply:
Yes. Colossalai is not designed to frequently set up and detroy the distributed environment, it assumed that users only need to initialize the distributed envionment in the start of their scripts and exit the distributed environment in the end. It keeps a global distributed configuration and ddp.teardown will damage this configuration.